Typescript sees the Boolean Constructor Type as Date when using Mongoose

### Prerequisites

- [X] I have written a descriptive issue title
- [X] I have searched existing issues to ensure the bug has not already been reported


### Mongoose version

8.4.0

### Node.js version

21.7.1

### MongoDB server version

6.6.2

### Typescript version (if applicable)

5.4.5

### Description

Whenever I try to make an Schema with a property of the Type Boolean Constructor the typescript server sees it as being of the type Date.

 error TS2322: Type 'boolean' is not assignable to type 'Date'.

22     course.isPublished =  false;
       ~~~~~~~~~~~~~~~~~~


Found 1 error in src/updatemongo.ts:22

### Steps to Reproduce

I just create a TypeScript project with this tsconfig

{
  "compilerOptions": {
    "target": "ES2015",
    "module": "commonjs",
    "esModuleInterop": true,
    "moduleResolution": "Node",                             
    "forceConsistentCasingInFileNames": true,            
    "strict": true,                                      
    "skipLibCheck": true,
    "rootDir": "src",
    "outDir": "dist"                              
  }
}

install the latest mongoose available using npm and 
try to make an Scheme like this.

import mongoose, { Mongoose } from "mongoose";
const courseSchema = new mongoose.Schema({
    name: String,
    author: String,
    tags: [String],
    date: { type: Date, default: Date.now },
    prices:Number,
    isPublished: Boolean
});

const Course = mongoose.model('Course', courseSchema);

async function updateCourse(id:any) {
    const course = await Course.findById(id);
    if (!course) return;
    course.isPublished =  false; //tsc won't compile because it will see it as being of type Date
    course.author = 'Another Author';
}

updateCourse("none");

<img width="334" alt="BooleanASDate" src="https://github.com/Automattic/mongoose/assets/39200671/e615cea2-9bcb-4407-9fe5-46ebd3b74465">


### Expected Behavior

Typescript should see it as being of type Boolean

> I'm unable to repro. The following script compiles correctly with Mongoose 8.4.0 and TypeScript 5.4.5. Please modify the following script to demonstrate your issue. Also, just to be clear, do you get this error when you compile TypeScript, or does this error just show up when you hover over `isPublished` in your code editor?
> 
> ```ts
> import mongoose, { Mongoose } from "mongoose";
> const courseSchema = new mongoose.Schema({
>   name: String,
>   author: String,
>   tags: [String],
>   date: { type: Date, default: Date.now },
>   prices:Number,
>   isPublished: Boolean
> });
> 
> const Course = mongoose.model('Course', courseSchema);
> 
> async function updateCourse(id:any) {
>   const course = await Course.findById(id);
>   if (!course) return;
>   course.isPublished = false; //tsc won't compile because it will see it as being of type Date
>   course.author = 'Another Author';
> }
> 
> updateCourse('none');
> ```
> 
> Output:
> 
> ```
> $ ./node_modules/.bin/tsc --strict --target ES2015 --moduleResolution node --esModuleInterop gh-14630.ts 
> $ head node_modules/typescript/package.json 
> {
>     "name": "typescript",
>     "author": "Microsoft Corp.",
>     "homepage": "https://www.typescriptlang.org/",
>     "version": "5.4.5",
>     "license": "Apache-2.0",
>     "description": "TypeScript is a language for application scale JavaScript development",
>     "keywords": [
>         "TypeScript",
>         "Microsoft",
> $ 
> ```

This is the full code I have and not only the short snippet to get to the error 

updatemongo.ts

`import mongoose, { Mongoose } from "mongoose";

const uri = 'mongodb+srv://<user>:<password>@<mongo>/mongo-exercises?retryWrites=true&w=majority&appName=Molina1';
mongoose.connect(uri)
    .then(() => console.log('Connected to Mongo Database...'))
    .catch(err => console.error('Could not connect to Mongo Database...', err));

const courseSchema = new mongoose.Schema({
    name: String,
    author: String,
    tags: [String],
    date: { type: Date, default: Date.now },
    prices:Number,
    isPublished: Boolean
});

const Course = mongoose.model('Course', courseSchema);

async function updateCourse(id:any) {
    const course = await Course.findById(id);
    if (!course) return;
    course.isPublished =  false as any;
    course.author = 'Another Author';
    course.isPublished = true;

    course.set({
        isPublished:true,
        author:'Another Author'
    }); 

    const result = await course.save();
    console.log(result);
} 

/*async function updateCourse(id:any) {
    const course = await Course.updateOne({_id:id},{
        $set:{
            author:'Mosh',
            isPublished:true,
        },
        $currentDate:{
            date:true
        }
    });
    
    console.log(course);
}*/

updateCourse("665a0ee68b40030cbec8dd57");`

Both the code editor and the tsc command line would tell me that the data type is of type Date when using TypeScript. 
This is the message  I get from tsc

```shell
tsc
src/updatemongo.ts:24:5 - error TS2322: Type 'boolean' is not assignable to type 'Date'.

24     course.isPublished = true;
       ~~~~~~~~~~~~~~~~~~

Found 1 error in src/updatemongo.ts:24
``` 
And this is the error I get on the IDE from the typescript server:
![image](https://github.com/Automattic/mongoose/assets/39200671/c70f90f4-93c7-4c65-bc13-0d5eb36b4101)

 Let me paste the package.json and the package-lock.json, and the tsconfig.json
It may help

[package.json](https://github.com/user-attachments/files/15536160/package.json)
[package-lock.json](https://github.com/user-attachments/files/15536161/package-lock.json)
[tsconfig.json](https://github.com/user-attachments/files/15536162/tsconfig.json)


And I am sure I am on the versions of mongo, tsc and mongoose I placed at the start of the ticket.
If there is  other info I can give you other than the one I just placed here, please let me know.

--- a/types/index.d.ts
+++ b/types/index.d.ts
@@ -493,7 +493,7 @@ declare module 'mongoose' {
   export type NumberSchemaDefinition = typeof Number | 'number' | 'Number' | typeof Schema.Types.Number;
   export type StringSchemaDefinition = typeof String | 'string' | 'String' | typeof Schema.Types.String;
   export type BooleanSchemaDefinition = typeof Boolean | 'boolean' | 'Boolean' | typeof Schema.Types.Boolean;
-  export type DateSchemaDefinition = typeof NativeDate | 'date' | 'Date' | typeof Schema.Types.Date;
+  export type DateSchemaDefinition = DateConstructor | 'date' | 'Date' | typeof Schema.Types.Date;
   export type ObjectIdSchemaDefinition = 'ObjectId' | 'ObjectID' | typeof Schema.Types.ObjectId;
 
   export type SchemaDefinitionWithBuiltInClass<T> = T extends number

